### PR TITLE
Repository: fix commit tags being seen in data

### PR DESCRIPTION
**Pull request by enkore**
I have a slightly different fix for master that doesn't destroy performance with larger segments (i.e. put commits in their own segment)

Problem description: is_committed_segment() seeked to the end of a segment file and checked whether there is a COMMIT_TAG there. This can also be the case when a PUT ends with a structure that is a COMMIT_TAG and is thus no valid way to check for the commit state of a segment.

This is a luck-based bug. You'll need to back something up where a chunk ends with a commit tag (another borg repo) and then that has to end up at the end of a segment file and *then* you'll need a kill/crash and *then* you *can theoretically* get an inconsistent repo (with some other conditions applying as well).

Now:

- Early segment magic check by iterator setup in iter_objects
- Sanity check: Do we have a COMMIT_TAG at the end? (<-- this is also the fast-bail-out for practically all uncomitted segments)
- If no: can't be committed
- If yes: iterate objects, check if that's an actual commit tag

### CODE REVIEW

File: borg/repository.py [hunk 1]
```diff
@@ -550,18 +550,37 @@ def cleanup(self, transaction_id):
...
+                tag, key, offset = next(iterator)
+            except IntegrityError:
+                return False
+            except StopIteration:
+                break
+            if tag == TAG_COMMIT:
+                seen_commit = True
+                continue
+            if seen_commit:
+                raise IntegrityError('Commit not located at end of segment [segment {}, offset {}]'.format(segment, offset))
```
Comment (ThomasWaldmann) [inline borg/repository.py]: the old code would just return False in this case, not raise.

File: borg/repository.py [hunk 2]
```diff
@@ -582,6 +601,8 @@ def get_write_fd(self, no_new=False, raise_full=False):
         return self._write_fd
 
     def get_fd(self, segment):
+        if segment == self.segment and self._write_fd:
+            self._write_fd.flush()
```
Comment (ThomasWaldmann) [inline borg/repository.py]: this seems a unrelated fix.

File: borg/repository.py [hunk 3]
```diff
@@ -550,18 +550,37 @@ def cleanup(self, transaction_id):
...
+        with open(self.segment_filename(segment), 'rb') as fd:
             try:
                 fd.seek(-self.header_fmt.size, os.SEEK_END)
             except OSError as e:
                 # return False if segment file is empty or too small
                 if e.errno == errno.EINVAL:
                     return False
                 raise e
-            return fd.read(self.header_fmt.size) == self.COMMIT
+            if fd.read(self.header_fmt.size) != self.COMMIT:
```
Comment (enkore) [inline borg/repository.py]: I wonder why write_commit isn't using this handy constant

File: borg/testsuite/repository.py [hunk 4]
```diff
@@ -268,7 +275,7 @@ def list_objects(self):
         return set(int(key) for key in self.repository.list())
 
     def test_repair_corrupted_segment(self):
-        self.add_objects([[1, 2, 3], [4, 5, 6]])
+        self.add_objects([[1, 2, 3], [4, 5], [6]])
```
Comment (ThomasWaldmann) [inline borg/testsuite/repository.py]: why did you change this?

**Comment by ThomasWaldmann** [comment]
besides unclear test change, code looks good to me.

**Comment by enkore** [comment]
IIRC there is a check later that accessing an object in the last segment raises IntegrityError. This doesn't happen anymore with this code, because the last segment won't be considered committed anymore due to the IntegrityError in it.

**Comment by ThomasWaldmann** [comment]
ok, merge or is something left to do?

**Comment by enkore** [comment]
I will make another PR about this for master after we back-merge 1.0-maint to fix a performance nag (not a bug) this code has for large segments. That involves much more changes to the tests hence nothing for 1.0-maint (which wouldn't be used with large segments).

**Comment by enkore** [comment]
This was also reported a couple years back in attic. At that time it didn't support uncompressed storage, so it wasn't an issue: https://github.com/jborg/attic/issues/80

**Comment by ThomasWaldmann** [comment]
@enkore maybe rather add this at another place than a closed pull request?
